File: javascript/hackerrank/minimum-loss.js
'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.replace(/\s*$/, '')
        .split('\n')
        .map(str => str.replace(/\s*$/, ''));

    main();
});

function readLine() {
    return inputString[currentLine++];
}

// Complete the minimumLoss function below.
function minimumLoss(price) {
    let sorted = price.map(function(v, i) {
        return [i, v]
    })
    sorted = sorted.sort(function(a, b) {
        if (a[1] < b[1]) {
            return 1
        } else if (a[1] > b[1]) {
            return -1
        } else {
            return 0
        }
    })
    let minLoss = Infinity;
    for (var i = 1; i < sorted.length; i++) {
        if (sorted[i - 1][0] < sorted[i][0] && sorted[i - 1][1] - sorted[i][1] < minLoss) {
            minLoss = sorted[i - 1][1] - sorted[i][1]
        }
    }
    return minLoss
}

function main() {
    const ws = fs.createWriteStream(process.env.OUTPUT_PATH);

    const n = parseInt(readLine(), 10);

    const price = readLine().split(' ').map(priceTemp => parseInt(priceTemp, 10));

    let result = minimumLoss(price);

    ws.write(result + "\n");

    ws.end();
}

File: python/hackerrank/minimum-loss.py
#!/bin/python3

import math
import os
import random
import re
import sys

# Complete the minimumLoss function below.
def minimumLoss(price):
	sp = [[i, v] for i, v in enumerate(price)]
	sp = sorted(sp, key=lambda p: p[1], reverse=True)
	min_loss = None
	for i in range(1, len(sp)):
		if min_loss is None or (sp[i - 1][0] < sp[i][0] and sp[i - 1][1] - sp[i][1] < min_loss):
			min_loss = sp[i - 1][1] - sp[i][1]
	return min_loss

if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    n = int(input())

    price = list(map(int, input().rstrip().split()))

    result = minimumLoss(price)

    fptr.write(str(result) + '\n')

    fptr.close()
